### PR TITLE
Feature/4133 always enable pagespeed insights

### PR DESCRIPTION
**Pull request by ivankruchkoff**
## Summary

Addresses issue #4133

## Relevant technical choices

- One item not covered in IB was covered in subsequent comments https://github.com/google/site-kit-wp/issues/4133#issuecomment-956505945
- 
## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (ivankruchkoff) [inline assets/js/feature-tours/pagespeed-insights-force-active.js:39]: what version should we set this to?

It can't be `n.e.x.t.` as we get a non semver error.

File: includes/Core/Assets/Assets.php
Comment (eugene-manuilov) [inline includes/Core/Assets/Assets.php:687]: Instead of adding this information here, let's better update the appropriate hook in the Modules class: https://github.com/google/site-kit-wp/blob/740fa3262c465dafd097a7b8d4100bf68c6a52c4/includes/Core/Modules/Modules.php#L247-L262

File: includes/Core/Util/Migration_1_45_0.php
Comment (eugene-manuilov) [inline includes/Core/Util/Migration_1_45_0.php:38]: > A potential way to solve this is could be:
> * Include an upgrade routine that sets a very simple option in the DB like googlesitekit_psi_updated if the site is currently connected.
> * **Show the tooltip tour only if that flag is set (and if PSI is not in the active modules option).**
> * On reset, delete this option so that it's gone.

We also need to check this flag when we determine a value for the `psiManuallyActivated` property.

File: includes/Modules/PageSpeed_Insights.php
Comment (eugene-manuilov) [inline includes/Modules/PageSpeed_Insights.php:20]: Is this needed here?

File: includes/Core/Util/Migration_1_45_0.php
Comment (eugene-manuilov) [inline includes/Core/Util/Migration_1_45_0.php:32]: We need to instantiate this class in the Plugins class as we do it for other migration classes: https://github.com/google/site-kit-wp/blob/740fa3262c465dafd097a7b8d4100bf68c6a52c4/includes/Plugin.php#L194-L195

File: includes/Core/Util/Migration_1_45_0.php
Comment (eugene-manuilov) [inline includes/Core/Util/Migration_1_45_0.php:43]: All `@since` phpdoc tags in this file should be `n.e.x.t` as well as the `DB_VERSION` constant value.

File: includes/Core/Util/Migration_1_45_0.php
Comment (eugene-manuilov) [inline includes/Core/Util/Migration_1_45_0.php:3]: I think it should be `1.46.0`, shouldn't it? The whole file and the class name should be renamed to `Migration_1_46_0`.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @ivankruchkoff. This is a good start. Added some comments for you. Please, take a look.

File: includes/Core/Util/Migration_1_45_0.php
Comment (ivankruchkoff) [reply includes/Core/Util/Migration_1_45_0.php]: https://github.com/google/site-kit-wp/pull/4305/commits/95aec903472e378b824d99b9e3a98f9b390d8a3b

File: assets/js/modules/pagespeed-insights/components/dashboard/DashboardPageSpeedWidget.js
Comment (ivankruchkoff) [reply assets/js/modules/pagespeed-insights/components/dashboard/DashboardPageSpeedWidget.js:41]: https://github.com/google/site-kit-wp/pull/4305/commits/7ed480fe70a177cc0254af2c556f68528838c3a4

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (ivankruchkoff) [reply assets/js/feature-tours/pagespeed-insights-force-active.js]: https://github.com/google/site-kit-wp/pull/4305/commits/72b81242ce456373491ec98d81bcbb2c46391b68

File: includes/Modules/PageSpeed_Insights.php
Comment (ivankruchkoff) [reply includes/Modules/PageSpeed_Insights.php]: https://github.com/google/site-kit-wp/pull/4305/commits/999698b4485ce25311ef90750c157566e1e7ada6

File: includes/Core/Assets/Assets.php
Comment (ivankruchkoff) [reply includes/Core/Assets/Assets.php]: https://github.com/google/site-kit-wp/pull/4305/commits/a76c3de66869fc93743b5ac7b48a1bdf044d7d26

File: includes/Core/Assets/Assets.php
Comment (eugene-manuilov) [inline includes/Core/Assets/Assets.php:14]: Changes in this file aren't needed. Please, remove it.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @ivankruchkoff. This is closer now, but still require a bit of work. Please, see my comment below and two more comments from the previous review: https://github.com/google/site-kit-wp/pull/4305#discussion_r745055355 and https://github.com/google/site-kit-wp/pull/4305#discussion_r745057502

File: includes/Core/Assets/Assets.php
Comment (ivankruchkoff) [reply includes/Core/Assets/Assets.php]: https://github.com/google/site-kit-wp/pull/4305/commits/9b103db948a1d0793a69c00688625e918fde282d (another unused import went with it).

File: includes/Core/Util/Migration_1_45_0.php
Comment (ivankruchkoff) [reply includes/Core/Util/Migration_1_45_0.php]: I thought I had done this... https://github.com/google/site-kit-wp/pull/4305/commits/50a92f41a8c0e6ac53fc8667ce803c739f6a0e37

File: includes/Core/Util/Migration_1_45_0.php
Comment (ivankruchkoff) [reply includes/Core/Util/Migration_1_45_0.php]: https://github.com/google/site-kit-wp/pull/4305/commits/43021fd8f08fdf722074790b62021424c4a7db18

File: includes/Core/Assets/Assets.php
Comment (ivankruchkoff) [reply includes/Core/Assets/Assets.php]: https://github.com/google/site-kit-wp/commit/9b103db948a1d0793a69c00688625e918fde282d

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (eugene-manuilov) [inline assets/js/feature-tours/pagespeed-insights-force-active.js:38]: I think this should be `1.46.0` since we are going to display this feature tour only for users who have installed the plugin earlier.

File: includes/Core/Util/Migration_1_46_0.php
Comment (eugene-manuilov) [inline includes/Core/Util/Migration_1_46_0.php:32]: I have been testing this on my end and it seems like we don't need this migration class at all. The tooltip tour won't re-appear after reset because dismissed feature tours are saved in the persistent option. Thus I think we can remove this class.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
@ivankruchkoff, added three more comments. Please, take a look.

File: includes/Core/Util/Migration_1_46_0.php
Comment (eugene-manuilov) [reply includes/Core/Util/Migration_1_46_0.php:32]: cc @felixarntz for visibility.

Comment (ivankruchkoff) [reply includes/Core/Util/Migration_1_46_0.php:32]: @eugene-manuilov yes, in that scenario it would be fine, but we're trying to avoid a different one:

https://github.com/google/site-kit-wp/issues/4133#issuecomment-958647430

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (ivankruchkoff) [reply assets/js/feature-tours/pagespeed-insights-force-active.js]: https://github.com/google/site-kit-wp/pull/4305/commits/e127eb1e08d6fc09cfb94ff99bde0f7761f56520

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (ivankruchkoff) [reply assets/js/feature-tours/pagespeed-insights-force-active.js]: https://github.com/google/site-kit-wp/pull/4305/commits/e127eb1e08d6fc09cfb94ff99bde0f7761f56520

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (felixarntz) [reply assets/js/feature-tours/pagespeed-insights-force-active.js]: @ivankruchkoff @eugene-manuilov Sounds good - we just need to ensure then that this actually ships with 1.46.0. I've therefore set this release on the issue to mark it already as "requirement".

Overall, where does the semver error come from? It would be great if we could specify `n.e.x.t` here as we do elsewhere - maybe we can allow for a special case (e.g. `n.e.x.t` could _always_ consider be greater than any other version you compare it too). Nothing to break our heads about here, but maybe worth a separate issue.

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (felixarntz) [inline assets/js/feature-tours/pagespeed-insights-force-active.js:54]: Is this really needed? We point to the overall widget area, so I don't think it matters for the tooltip where the widget is fully loaded. I'd say let's remove this, less code to maintain.

File: includes/Core/Modules/Modules.php
Comment (felixarntz) [inline includes/Core/Modules/Modules.php:481]: Let's not make this public - it's not a method intended for others to call, and it doesn't need to be public for the code here either.

File: includes/Core/Util/Migration_1_46_0.php
Comment (felixarntz) [reply includes/Core/Util/Migration_1_46_0.php:32]: @eugene-manuilov @ivankruchkoff is correct - this extra flag is needed to indicate that the plugin was already connected before. In combination with the check for whether the module is not manually activated, this tells us that:
* the site was using Site Kit before and was already connected
* the site didn't have PSI connected

Because of those conditions we need to show the tooltip tour. On reset, this option here will get wiped, which is intentional because after reset the site starts in a fresh state. In that situation, we shouldn't ever show the tooltip tour because activating PSI after setup is more of an expected scenario, while if the site already _was_ connected, it would be a more "random" experience where we have to explain why it's suddenly active.

File: includes/Core/Modules/Modules.php
Comment (felixarntz) [inline includes/Core/Modules/Modules.php:261]: I think the logic here is slightly off, in combination with https://github.com/google/site-kit-wp/pull/4305/files#diff-b0980b9f904136a052c4068889aa1973f11d6a7fc5d1c784bf37329ac8c13f79R56. For example, right now it will evaluate to `false` when a site that is just starting with 1.46.0 (i.e. no migration) hasn't manually activated PSI, which for that version will always be the case (since you can't activate it anymore).

This will lead to the tooltip tour potentially showing, which is not correct.

I suggest we simplify this: We don't actually care for whether PSI was manually activated. We care for whether PSI is **not** manually activated and the current version was migrated to from a pre-1.46.0 version. I think this here could simply become:
```suggestion
				$data['showPSIForceActiveTour'] = ! $this->manually_enabled( 'pagespeed-insights' ) && $this->options->get( Migration_1_46_0::OPTION_KEY_PSI_UPDATED );
```

Then https://github.com/google/site-kit-wp/pull/4305/files#diff-b0980b9f904136a052c4068889aa1973f11d6a7fc5d1c784bf37329ac8c13f79R56 should check for whether that value is `true`.

**Review by felixarntz — CHANGES_REQUESTED** [review]
@ivankruchkoff Left a few comments here. This mostly looks good, but I believe there's one critical logic problem.

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (ivankruchkoff) [reply assets/js/feature-tours/pagespeed-insights-force-active.js:54]: See my comment in https://github.com/google/site-kit-wp/issues/4133#issuecomment-965912725 for the issues with each approach.

File: includes/Core/Modules/Modules.php
Comment (ivankruchkoff) [reply includes/Core/Modules/Modules.php]: https://github.com/google/site-kit-wp/pull/4305/commits/6e07da780fa8d2be9e85a194dc55592b879d4b07

File: assets/js/feature-tours/pagespeed-insights-force-active.js
Comment (ivankruchkoff) [reply assets/js/feature-tours/pagespeed-insights-force-active.js]: https://github.com/google/site-kit-wp/pull/4305/commits/6e07da780fa8d2be9e85a194dc55592b879d4b07 addresses the CR comment.

File: includes/Core/Modules/Modules.php
Comment (ivankruchkoff) [reply includes/Core/Modules/Modules.php]: https://github.com/google/site-kit-wp/pull/4305/commits/96b968af5c7efc192fb6ddd4c5e3269ad0f1debe

File: includes/Core/Util/Migration_1_46_0.php
Comment (eugene-manuilov) [reply includes/Core/Util/Migration_1_46_0.php:32]: Thanks, @felixarntz and @ivankruchkoff.

File: includes/Core/Modules/Modules.php
Comment (eugene-manuilov) [inline includes/Core/Modules/Modules.php:261]: @ivankruchkoff, this line is not needed anymore, right?

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]

File: includes/Core/Modules/Modules.php
Comment (ivankruchkoff) [reply includes/Core/Modules/Modules.php]: https://github.com/google/site-kit-wp/pull/4305/commits/04da3b7f7f3fdb28dea93771cf83620555d4c817

**Review by eugene-manuilov — APPROVED** [review]
Ok, this looks good to me. Thanks, @ivankruchkoff! @felixarntz leaving this for you to approve and merge if you have nothing more to add.

**Review by felixarntz — APPROVED** [review]
LGTM given that we go with the approach from https://github.com/google/site-kit-wp/issues/4133#issuecomment-966515572.

Before merging this, please review https://github.com/google/site-kit-wp/issues/4133#issuecomment-966598285 though - if we can find a solution like that, let's adjust.

**Review by felixarntz — APPROVED** [review]

**Pull request merged**
